estimagic/optimization/process_arguments.py

```diff
@@ -0,0 +1,311 @@
+            res = argument[0]
+        else:
+            res = argument
+
+    # Argument is broadcasted
+    elif len_arg == 1:
+
+        # Select element if entered as list of length 1
+        if isinstance(argument, (list, tuple)) and len(argument) == 1:
+            res = [copy.deepcopy(argument[0]) for i in range(n_opts_total)]
+        else:
+            res = [copy.deepcopy(argument) for i in range(n_opts_total)]
+
+    # Argument was entered as list/tuple
+    elif len_arg == n_opts_total:
+        res = argument
+
+    # Argument was entered as too short list/tuple
+    else:
+        raise ValueError(msg)
+    return res
+
+
+def process_optimization_arguments(
+    criterion,
+    params,
+    algorithm,
+    criterion_kwargs=None,
+    constraints=None,
+    general_options=None,
+    algo_options=None,
+    dashboard=False,
+    db_options=None,
+):
+    """Process and validate arguments for minimize or maximize.
+
+    Args:
+        criterion (function or list of functions):
+            Python function that takes a pandas Series with parameters as the first
+            argument and returns a scalar floating point value.
+
+        params (pd.DataFrame or list of pd.DataFrames):
+            See :ref:`params`.
+
+        algorithm (str or list of strings):
+            specifies the optimization algorithm. See :ref:`list_of_algorithms`.
+
+        criterion_kwargs (dict or list of dicts):
+            additional keyword arguments for criterion
+
+        constraints (list or list of lists):
+            list with constraint dictionaries. See for details.
+
+        general_options (dict):
+            additional configurations for the optimization
+
+        algo_options (dict or list of dicts):
+            algorithm specific configurations for the optimization
+
+        dashboard (bool):
+            whether to create and show a dashboard
+
+        db_options (dict):
+            dictionary with kwargs to be supplied to the run_server function.
+    """
+
+    # set default arguments
+    criterion_kwargs = {} if criterion_kwargs is None else criterion_kwargs
+    constraints = [] if constraints is None else constraints
+    algo_options = {} if algo_options is None else algo_options
+    db_options = {} if db_options is None else db_options
+    general_options = {} if general_options is None else general_options
+
+    # Determine number of optimizations and check types
+
+    # Specify name and expected type for all arguments.
+    # Three groups of arguments are relevant:
+    # (i) expected as scalar,
+    # (ii) expected as scalar or list/tuple,
+    # (iii) expected as list/tuple or nested list/tuple
+
+    args_non_list = [
+        {"candidate": general_options, "scalar_type": dict, "name": "general_options"},
+        {"candidate": dashboard, "scalar_type": bool, "name": "dashboard"},
+        {"candidate": db_options, "scalar_type": dict, "name": "db_options"},
+    ]
+    args_list = [
```

Comment:
From here until line 279 the arguments are handled and matched based on their order in some list. Then in 299+ they are put back into a dictionary. 

The position based handling is error prone and in fact there was a mistake in the order of `algo_options` and `criterion_kwargs` that the `test_optimize.py` (merged from the master), discovered. I fixed it for now, but it would be nicer to always handle the arguments based on names, i.e. instead of `args_list` you would have a dictionary where the keys are the names and the values are dictionaries that contain the remaining information, including an entry called "type" or "kind" that contains the information that you currently encode by splitting the arguments into several lists.